### ISSUE TITLE
Accelerate Multi-GPU set-up unable to send tensors to cpu() after calling gather()

### ISSUE THREAD

**Issue by nils-hde** [huggingface/accelerate#1688]
### System Info

```Shell
Using latest transformers accelerate releases with PyTorch 1.13.1 and this code snipped
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

`    # Gather the tensors
    logger.info("Gathering results from all GPUs")
    all_sampled_outputs = accelerator.gather(all_sampled_outputs)
    all_ground_truths = accelerator.gather(all_ground_truths)
    result = dict()

    # Remove padding and convert back to lists
    logger.info(f"Gathered results from all GPUs. Preparing results for decoding.")
    logger.info(f"Shape of gathered GTs: {all_ground_truths.shape}")
    logger.info(f"Shape of gathered predictions: {all_sampled_outputs.shape}")
    all_sampled_outputs = all_sampled_outputs.cpu()
    all_ground_truths = all_ground_truths.cpu()
    logger.info("Back on CPU")
    all_sampled_outputs = all_sampled_outputs.tolist()
    all_ground_truths = all_ground_truths.tolist()
    logger.info("Converted tensors back to lists on CPU. Continuing with decoding")

    if accelerator.is_main_process:
        all_sampled_texts = [[tokenizer.decode(sampled_output, skip_special_tokens=True) for sampled_output in sampled_outputs] for sampled_outputs in all_sampled_outputs]
        all_ground_truths_text = [tokenizer.decode(ground_truth, skip_special_tokens=True) for ground_truth in all_ground_truths]

        logger.info("Finished decoding. Continuing with text stripping and writing the predictions.")
        
        # Remove leading white spaces if necessary
        all_sampled_texts = [[text.lstrip() for text in texts] for texts in all_sampled_texts]
        all_ground_truths_text = [text.lstrip() for text in all_ground_truths_text]`

all_sampled_outputs and all_ground_truths are 2D tensors with only about 30 entries in each dimension. Still the code never passes ".cpu()" stage and runs forever on any multiGPU setup with this config for accelerate

```
compute_environment: LOCAL_MACHINE
deepspeed_config:
  deepspeed_config_file: ds_config.json
  zero3_init_flag: true
distributed_type: DEEPSPEED
downcast_bf16: 'no'
dynamo_config:
  dynamo_backend: 'no'
machine_rank: 0
main_training_function: main
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false

```

### Expected behavior

Tried moving the code piece out of the main_process part, no difference. It simply seems to be impossible to move the gathered tensor to the CPU. Code runs without problem when there is no distributed setup. Any ideas?

**Comment by nils-hde** [comment]
I saw another issue and found the problem is that I haven’t padded „across processes“ but only using the torch padding function which lead to different tensor lengths. Afterwards no more problems

**Comment by muellerzr** [comment]
Great! We're working on a debugging system that can help with situations like this as well here soon to make it less frustrating, but nice investigative work

**Comment by ryan-caesar-ramos** [comment]
Hi! Can I ask how distributed evaluation works with `pad_across_processes`?

If I'm not mistaken, the distributed evaluation part of the quick tour docs states that

> Data at the end of a dataset may be duplicated so the batch can be equally divided among all workers. The [gather_for_metrics()](https://huggingface.co/docs/accelerate/v0.28.0/en/package_reference/accelerator#accelerate.Accelerator.gather_for_metrics) method automatically removes the duplicated data to calculate a more accurate metric.

However I ran into the same problem originally stated where my code would hang after running `gather_for_metrics` and before moving tensors to CPU. I found that after gathering, the sizes seemed to be inconsistent, which I assume points to my original tensors having inconsistent sizes.

Firstly, I'm not sure how to get this automatic duplication to kick in (I'm assuming it's not, and if it were I would be having different errors). Secondly, even if it did kick in, am I correct in understanding that the de-duplication is based off of gradient information? At least that's how I understood it based off the source code for `gather_with_metrics`. Since I'm just running inference for metrics/evaluation purposes, everything's under the `torch.no_grad()` context wrapper, so I'm not sure if that throws a wrench in things?

Eventually I used `pad_across_processes` before `gather_for_metrics`, and my main question is: will my evaluations come out alright? Will the padded data be accounted for after gathering or do I need to handle that myself?

Thanks!
